R/checkPythonDeps.R

```diff
@@ -0,0 +1,350 @@
+    ))
+  } else if (style == "conda") {
+    # Example: "pip=24.0=pyhd8ed1ab_0" -> Split at "=", first part is dependency name, second part is version,
+    # third part is build
+    splitDepString <- strsplit(depString, "=")[[1]]
+    extractPythonVersion(splitDepString[2])
+    return(createPythonDependency(
+      splitDepString[1],
+      "==",
+      extractPythonVersion(splitDepString[2]),
+      build = splitDepString[3],
+      repo = ""
+    ))
+  } else if (style == "pip") {
+    # Example: "climate_assessment==0.1.4a0" -> Split at "==", first part is dependency name, second part is version
+    pattern <- "(.*)(==|>=|<=|>|<|!=)(.*)"
+    splitDepString <- strcapture(pattern, depString, proto = list(name = "", operator = "", version = ""))
+    return(createPythonDependency(
+      splitDepString$name,
+      splitDepString$operator,
+      extractPythonVersion(splitDepString$version),
+      build = "",
+      repo = ""
+    ))
+  } else {
+    stop("Invalid argument: 'style' should be either 'pip' or 'conda'")
+  }
+}
+
+# ---------------------
+# Main Functions
+# ---------------------
+
+#' Check Python Dependencies and Optionally Their Versions
+#'
+#' Checks if the required Python dependencies can actually be imported in the Python environment provided. Relies on
+#' reticulate::import. Optionally check if versions are met
+#'
+#' @param dependencies Vector of dependency strings
+#' @param action Action to take if a dependency is missing. Either "stop", "warn", "note", or "pass"
+#' @param strict Logical indicating whether to check for matching versions
+#' @return TRUE if all dependencies are installed, FALSE otherwise
+#' @examples
+#' \dontrun{
+#' deps <- c("climate_assessment==0.1.4a0", "numpy<2.0")
+#' checkPythonDeps(deps, action="stop", strict = TRUE)
+#' }
+#' @export
+checkPythonDeps <- function(dependencies, action = "stop", strict = TRUE) {
+  if (!action %in% c("stop", "warn", "note", "pass")) {
+    stop("Invalid action '", action, "'. Must be one of 'stop', 'warn', 'note' or 'pass'.")
+  }
+  # Keep track of missing dependencies
+  missingDependencies <- c()
+  for (dependencyString in dependencies) {
+    dependency <- extractPythonDependency(dependencyString)
+    tryCatch(
+      {
+        pythonModule <- import(dependency$name)
+        if (strict) {
+          installedVersion <- extractPythonVersion(pythonModule$`__version__`)
+          correctVersion <- comparePythonVersions(dependency$operator, installedVersion, dependency$version)
+          if (!correctVersion) {
+            missingDependencies <- c(
+              missingDependencies,
+              paste0(dependencyString, " (found version: ", printPythonVersion(installedVersion), ")")
+            )
+          }
+        }
+      },
+      error = function(e) {
+        # Need to use <<- here to assign to the variable in the parent environment
+        missingDependencies <<- c(missingDependencies, paste0(dependencyString, " (missing) ")) # nolint
```

Comment:
`<<-` should really be avoided if at all possible. Is it possible to return what you want to add here and then add it in a scope where missingDependencies is accessible?

Reply:
I was not able to find another way to access `missingDependencies` form the error callback. I'd argue that this is the exception to the (otherwise perfectly fine) rule to avoid `<<-`, as the error case is two-fold: We're not only interested in dependencies that cannot be imported (which would trigger the error callback) but also a version mismatch (strict mode), where the dependency *can* be imported but has a version different from what's listed in the requirements. 

I wanna respect your best R-practices here: If you strictly oppose `<<-`-usage, I'll isolate the whole try-catch-block in a separate function that returns NULL if everything's fine, else returns a string giving a reason

Reply:
I do strictly oppose `<<-`, but I want to stress this is not "my" best R-practice :wink:

Reply:
Imma make it a function then :)